pkg/datapath/linux/ipsec/ipsec_linux_test.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build privileged_tests
+
+package ipsec
+
+import (
+	"net"
+	"os"
+	"testing"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/vishvananda/netlink"
+)
+
+// Hook up gocheck into the "go test" runner.
+func Test(t *testing.T) { TestingT(t) }
+
+type IPSecSuitePrivileged struct{}
+
+var _ = Suite(&IPSecSuitePrivileged{})
+
+var (
+	path             = "ipsec_keys_test"
+	dfltKeysDat    = []byte("hmac(sha256) abcdefghijklmnopqrstuvwzyzABCDEF cbc(aes) abcdefghijklmnopqrstuvwzyzABCDEF\n")
+	scopedKeysDat  = []byte("hmac(sha256) abcdefghijklmnopqrstuvwzyzABCDEF cbc(aes) abcdefghijklmnopqrstuvwzyzABCDEF foobar\n")
+	invalidKeysDat = []byte("test abcdefghijklmnopqrstuvwzyzABCDEF test abcdefghijklmnopqrstuvwzyzABCDEF\n")
+)
+
+func (p *IPSecSuitePrivileged) TestLoadKeysNoFile(c *C) {
+	err := LoadIPSecKeysFile(path)
+	c.Assert(err, ErrorMatches, "failed to load IPSec Keys File ipsec_keys_test: open ipsec_keys_test: no such file or directory")
+}
+
+func (p *IPSecSuitePrivileged) TestInvalidLoadKesyFile(c *C) {
+	f, err := os.Create(path)
+	c.Assert(err, IsNil)
+	defer f.Close()
+	_, err = f.Write(invalidKeysDat)
+	c.Assert(err, IsNil)
+	f.Sync()
+	err = LoadIPSecKeysFile(path)
```

Comment:
If `LoadIPSecKeysFile()` takes an `io.Writer` instead of a path (or there's a private variation on the function that does so), then you wouldn't need to write things to the filesystem to test this functionality.

Reply:
This is a nice idea from the ipsec_linux.go side we can abstract the file out of a base routine so we get, `func loadIPSecKeys(r io.Reader) error`  and then `LoadIPSecKeysFile` opens the file and passes the io.Reader into `loadIPSecKeys`. This makes the unit test independent of the file system we just need to create the "keys" buffer but also it means in the future we can pull keys easily from any io.Reader not just files easily.